lib/active_merchant/billing/gateways/hi_pay.rb

```diff
@@ -0,0 +1,203 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class HiPayGateway < Gateway
+      self.test_url = 'https://stage-secure-gateway.hipay-tpp.com/rest'
+      self.live_url = 'https://secure-gateway.hipay-tpp.com/rest'
+
+      self.supported_countries = %w[FR]
+      self.default_currency = 'EUR'
+      self.money_format = :cents
+      self.supported_cardtypes = %i[visa master american_express]
+
+      self.homepage_url = 'https://hipay.com/'
+      self.display_name = 'HiPay'
+
+      def initialize(options = {})
+        requires!(options, :username, :password)
+        @username = options[:username]
+        @password = options[:password]
+        super
+      end
+
+      def purchase(money, payment_method, options = {})
+        authorize(money, payment_method, options.merge({ operation: 'Sale' }))
+      end
+
+      def authorize(money, payment_method, options = {})
+        MultiResponse.run do |r|
+          if payment_method.is_a?(CreditCard)
+            response = r.process { tokenize(payment_method, options) }
+            card_token = response.params['token']
+          elsif payment_method.is_a?(String)
+            card_token, payment_product = payment_method.split('|')
+          end
+
+          post = {}
+          post[:payment_product] = payment_product&.downcase || payment_method.brand
+
+          post[:operation] = options[:operation] || 'Authorization'
+          add_product_data(post, options)
+          add_invoice(post, money, options)
+          post[:cardtoken] = card_token
+          r.process { commit('order', post) }
+        end
+      end
+
+      def capture(money, authorization, options)
+        post = {}
+        post[:operation] = 'capture'
+        post[:currency] = (options[:currency] || currency(money))
+
+        commit('capture', post, { transaction_reference: authorization })
+      end
+
+      def store(payment_method, options = {})
+        tokenize(payment_method, options.merge({ multiuse: '1' }))
+      end
+
+      def scrub(transcrip)
+        # code
+      end
+
+      private
+
+      def add_product_data(post, options)
+        post[:orderid] = options[:order_id] if options[:order_id]
+        post[:description] = options[:description]
+
+        # optional
+        # basket=> refer https://support.hipay.com/hc/fr/articles/115001660469-Payment-Gateway-Shopping-cart-management
+        post[:basket] = options[:basket] if options[:basket]
```

Comment:
❓ Question:

Any reason to add these fields to the initial implementation, are they required ?, they looks like Level 2 - 3 data

Reply:
I was not 100% sure, there are several I discarded but, I have no idea about those.